platform/security/sts/security-sts-realm/src/main/java/ddf/security/realm/sts/SamlRealm.java

```diff
@@ -0,0 +1,187 @@
+
+  private SamlAssertionValidator samlAssertionValidator;
+
+  private List<String> usernameAttributeList;
+
+  public SamlRealm() {
+    setCredentialsMatcher(new STSCredentialsMatcher());
+  }
+
+  /** Determine if the supplied token is supported by this realm. */
+  @Override
+  public boolean supports(AuthenticationToken token) {
+    boolean supported =
+        token != null && token.getCredentials() != null && token instanceof SAMLAuthenticationToken;
+
+    if (supported) {
+      LOGGER.debug("Token {} is supported by {}.", token.getClass(), SamlRealm.class.getName());
+    } else if (token != null) {
+      LOGGER.debug("Token {} is not supported by {}.", token.getClass(), SamlRealm.class.getName());
+    } else {
+      LOGGER.debug("The supplied authentication token is null. Sending back not supported.");
+    }
+
+    return supported;
+  }
+
+  /** Perform authentication based on the supplied token. */
+  @Override
+  protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token) {
+    Object credential = null;
+
+    // perform validation
+    if (token instanceof SAMLAuthenticationToken) {
+      try {
+        samlAssertionValidator.validate((SAMLAuthenticationToken) token);
+        credential = token.getCredentials();
+      } catch (AuthenticationFailureException e) {
+        String msg = "Unable to validate request's authentication.";
+        LOGGER.info(msg);
+        throw new AuthenticationException(msg, e);
+      }
+    }
+
+    if (credential == null) {
+      String msg =
+          "Unable to authenticate credential.  A NULL credential was provided in the supplied authentication token. This may be due to an error with the SSO server that created the token.";
+      LOGGER.info(msg);
+      throw new AuthenticationException(msg);
+    } else {
+      // removed the credentials from the log message for now, I don't think we should be dumping
+      // user/pass into log
+      LOGGER.debug("Received credentials.");
+    }
+
+    LOGGER.debug("Creating token authentication information with SAML.");
+    SimpleAuthenticationInfo simpleAuthenticationInfo = new SimpleAuthenticationInfo();
+    SecurityToken securityToken = checkForSecurityToken(credential);
+    SimplePrincipalCollection principals = createPrincipalFromToken(securityToken);
+    simpleAuthenticationInfo.setPrincipals(principals);
+    simpleAuthenticationInfo.setCredentials(credential);
+
+    return simpleAuthenticationInfo;
+  }
+
+  private SecurityToken checkForSecurityToken(final Object credential) {
+    if (credential instanceof PrincipalCollection) {
+      Optional<SecurityAssertionSaml> assertionSamlOptional =
+          ((PrincipalCollection) credential)
+              .byType(SecurityAssertionSaml.class)
+              .stream()
+              .filter(sa -> sa.getToken() instanceof SecurityToken)
+              .findFirst();
+      if (assertionSamlOptional.isPresent()) {
```

Comment:
✏️ Could collapse this to
```java
return ((PrincipalCollection) credential)
  .byType(SecurityAssertionSaml.class)
  .stream()
  .filter(sa -> sa.getToken() instanceof SecurityToken)
  .map(SecurityAssertionSaml::getToken)
  .map(SecurityToken.class::cast)
  .findFirst()
  .orElse(null);
```
There are a few other ways to do it as well. Could flatMap the tokens and avoid doing the instanceOf filter the way it is now, for example.